hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/repair/om/FSORepairTool.java

```diff
@@ -0,0 +1,782 @@
+   */
+  private static String buildReachableParentKey(String fileOrDirKey) {
+    String[] keyParts = fileOrDirKey.split(OM_KEY_PREFIX);
+    // Should be /volID/bucketID/parentID/name
+    // The first part will be blank since key begins with a slash.
+    Preconditions.assertTrue(keyParts.length >= 4);
+    String volumeID = keyParts[1];
+    String bucketID = keyParts[2];
+    String parentID = keyParts[3];
+
+    return OM_KEY_PREFIX +
+        volumeID +
+        OM_KEY_PREFIX +
+        bucketID +
+        OM_KEY_PREFIX +
+        parentID;
+  }
+
+  private void openReachableDB() throws IOException {
+    File reachableDBFile = new File(new File(omDBPath).getParentFile(),
+        "reachable.db");
+    System.out.println("Creating database of reachable directories at " + reachableDBFile);
+    // Delete the DB from the last run if it exists.
+    if (reachableDBFile.exists()) {
+      FileUtils.deleteDirectory(reachableDBFile);
+    }
+    reachableDBPath = reachableDBFile.toString();
+    reachableDB = buildReachableRocksDB(reachableDBFile);
+  }
+
+  private RocksDatabase buildReachableRocksDB(File reachableDBFile) throws IOException {
+    DBProfile profile = new OzoneConfiguration().getEnum(HDDS_DB_PROFILE, HDDS_DEFAULT_DB_PROFILE);
+    Set<TableConfig> tableConfigs = new HashSet<>();
+
+    try {
+      tableConfigs.add(new TableConfig("default", profile.getColumnFamilyOptions()));
+
+      return RocksDatabase.open(reachableDBFile,
+          profile.getDBOptions(),
+          new ManagedWriteOptions(),
+          tableConfigs, false);
+    } finally {
+      for (TableConfig config : tableConfigs) {
+        config.close();
+      }
+    }
+  }
+
+  private void closeReachableDB() {
+    if (reachableDB != null) {
+      reachableDB.close();
+    }
+  }
+
+  private void dropReachableTableIfExists() throws IOException {
+    try {
+      List<byte[]>
+          availableCFs = reachableDB.listColumnFamiliesEmptyOptions(reachableDBPath);
+      boolean cfFound = false;
+      for (byte[] cfNameBytes: availableCFs) {
+        if (new String(cfNameBytes, UTF_8).equals(new String(REACHABLE_TABLE_BYTES, UTF_8))) {
+          cfFound = true;
+          break;
+        }
+      }
+
+      if (cfFound) {
+        reachableDB.dropColumnFamily(reachableCFHandle);
+      }
+    } catch (RocksDBException ex) {
+      throw new IOException(ex.getMessage(), ex);
+    } finally {
+      if (reachableCFHandle != null) {
+        reachableCFHandle.close();
+      }
+    }
+  }
+
+  private void createReachableTable() throws IOException {
+    reachableCFHandle = reachableDB.createColumnFamily(
+        new ColumnFamilyDescriptor(REACHABLE_TABLE_BYTES));
+  }
+
+  /**
+   * Define a Report to be created.
+   */
+  public static class Report {
+    private long reachableBytes;
+    private long reachableFiles;
+    private long reachableDirs;
+    private long unreachableBytes;
+    private long unreachableFiles;
+    private long unreachableDirs;
```

Comment:
Would be nice to wrap (bytes, files, dirs) in an object, and use one instance for reachable and unreachable, each.